### PR TITLE
Add Url::make_relative() as the inverse of Url::join()

### PR DESCRIPTION
**Pull request by sdroege**
This creates a relative URL from a base URL and another absolute URL.

Fixes https://github.com/servo/rust-url/issues/666

### CODE REVIEW

File: url/src/lib.rs [hunk 1]
```diff
@@ -349,11 +351,144 @@ impl Url {
...
+    ///
+    /// ```rust
+    /// use url::Url;
+    /// # use url::ParseError;
+    ///
+    /// # fn run() -> Result<(), ParseError> {
+    /// let base = Url::parse("https://example.net/a/b.html")?;
+    /// let url = Url::parse("https://example.net/a/c.png")?;
+    /// let relative = base.make_relative(&url);
+    /// assert_eq!(relative.as_deref(), Some("c.png"));
```
Comment (sdroege) [inline url/src/lib.rs:374]:
```
---- src/lib.rs - Url::make_relative (line 366) stdout ----
error[E0599]: no method named `as_deref` found for type `std::option::Option<std::string::String>` in the current scope
```

What's the minimum supported Rust version here? `Option::as_deref()` is stable since 1.40

**Comment by sdroege** [comment]
The clippy failures are all unrelated to this PR. I can submit another PR for them if you agree that all those should be fixed, or let me know which of the lints you'd like to disable (personally I'm not a fan of `clippy::manual-range-contains`...)

File: url/src/lib.rs (hunk 1 above)
Comment (djc) [reply url/src/lib.rs:374]: CI runs with 1.36.0.

Comment (sdroege) [reply url/src/lib.rs:374]: Do you want to keep that, and if so, which of the many patterns to solve this would you prefer? `relative.as_ref().map(|s| s.as_str())`? :)

**Comment by djc** [comment]
Yup, sorry about the clippy failures. Another PR would be nice, though a separate commit in this PR is also fine. I think we should fix all of them except the ones about the unit errors, because those would break backwards compatibility.

File: url/src/lib.rs (hunk 1 above)
Comment (djc) [reply url/src/lib.rs:374]: This one seems fine.

Comment (sdroege) [reply url/src/lib.rs:374]: Done, thanks!

**Comment by sdroege** [comment]
clippy fixes are in https://github.com/servo/rust-url/pull/669

**Comment by sdroege** [comment]
Rebased on top of #669

**Comment by valenting** [comment]
Thanks for this PR. It looks really nice.
Would you mind adding a couple of tests for the negative? ie? returning None?

**Comment by sdroege** [comment]
Sure, will add a few tests for that later or tomorrow! Thanks for the review

**Comment by sdroege** [comment]
Added tests for the 4 possible error cases.

**Comment by sdroege** [comment]
And rebased for the merge conflicts. @valenting anything else missing here?
